sdk/storage/azblob/internal/structuredmsg/structured_message.go

```diff
@@ -0,0 +1,369 @@
+	totalCRC64 uint64
+}
+
+// NewStructuredMessageWriter creates a new StructuredMessageWriter
+func NewStructuredMessageWriter(writer io.Writer) *StructuredMessageWriter {
+	return &StructuredMessageWriter{
+		writer:     writer,
+		crc64Table: shared.CRC64Table,
+		segmentNum: 0,
+		totalCRC64: 0,
+	}
+}
+
+// WriteHeader writes the message header
+func (w *StructuredMessageWriter) WriteHeader(numSegments uint32) error {
+	if numSegments == 0 || numSegments > MaxSegments {
+		return ErrInvalidSegmentCount
+	}
+
+	header := make([]byte, HeaderSize)
+	header[0] = MessageVersion
+	header[1] = FlagNone
+	binary.LittleEndian.PutUint16(header[2:], PropCRC64)
+	binary.LittleEndian.PutUint32(header[4:], numSegments)
+
+	_, err := w.writer.Write(header)
+	return err
+}
+
+// WriteSegment writes a segment with its data and CRC64
+func (w *StructuredMessageWriter) WriteSegment(data []byte) error {
+	if len(data) > MaxSegmentSize {
+		return ErrSegmentTooLarge
+	}
+
+	// Calculate segment CRC64
+	segmentCRC64 := crc64.Checksum(data, w.crc64Table)
+
+	// Write segment header
+	header := make([]byte, SegmentHeaderSize)
+	binary.LittleEndian.PutUint32(header[0:4], w.segmentNum)
+	binary.LittleEndian.PutUint32(header[4:8], uint32(len(data)))
+	binary.LittleEndian.PutUint64(header[8:16], segmentCRC64)
+
+	if _, err := w.writer.Write(header); err != nil {
+		return err
+	}
+
+	// Write segment data
+	if _, err := w.writer.Write(data); err != nil {
+		return err
+	}
+
+	// Update cumulative CRC64
+	w.totalCRC64 = crc64.Update(w.totalCRC64, w.crc64Table, data)
+	w.segmentNum++
+
+	return nil
+}
+
+// WriteTrailer writes the message trailer with cumulative CRC64
+func (w *StructuredMessageWriter) WriteTrailer() error {
+	trailer := make([]byte, TrailerSize)
+	trailer[0] = MessageVersion
+	trailer[1] = FlagNone
+	binary.LittleEndian.PutUint16(trailer[2:], PropCRC64)
+	binary.LittleEndian.PutUint64(trailer[4:], w.totalCRC64)
+	_, err := w.writer.Write(trailer)
+	return err
+}
+
+// GetTotalCRC64 returns the cumulative CRC64 of all segments written
+func (w *StructuredMessageWriter) GetTotalCRC64() uint64 {
+	return w.totalCRC64
+}
+
+// StructuredMessageReader decodes the binary message format
+type StructuredMessageReader struct {
+	reader     io.Reader
+	crc64Table *crc64.Table
+	header     *MessageHeader
+	segmentNum uint32
+	totalCRC64 uint64
+	closed     bool
+}
+
+// MessageHeader represents the message header
+type MessageHeader struct {
+	Version     uint8
+	Flags       uint8
+	Properties  uint16
+	NumSegments uint32
+}
+
+// NewStructuredMessageReader creates a new StructuredMessageReader
+func NewStructuredMessageReader(reader io.Reader) *StructuredMessageReader {
+	return &StructuredMessageReader{
+		reader:     reader,
+		crc64Table: shared.CRC64Table,
+		segmentNum: 0,
+		totalCRC64: 0,
+		closed:     false,
+	}
+}
+
+// ReadHeader reads and validates the message header
+func (r *StructuredMessageReader) ReadHeader() (*MessageHeader, error) {
```

Comment:
Do customers ever need to call this (and `ReadTrailer`), or can these not be exported?